addons/coupon/models/product_product.py

```diff
@@ -0,0 +1,16 @@
+from odoo import models, _
+from odoo.exceptions import ValidationError
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    def unlink(self):
+        '''
+        This methods prevents deleting Reward line product that is linked
+        with Coupon/Promotion
+        :return:
+        '''
+        coupon_prog = self.env['coupon.program'].with_context(active_test=False).search_count([('discount_line_product_id', 'in', self.ids)])
+        if coupon_prog > 0:
+            raise ValidationError(_("You cannot delete the product(s) that is linked with Coupon or Promotion program, Archive instead."))
```

Comment:
This translation needs to be added to the modules .pot file.

Reply:
Thank you for the review. I made the changes per request.